templates/boltzmann.js

```diff
@@ -575,7 +579,105 @@ function applyHeaders (headers = {}) {
   }
 }
 
-const applyXFO = (mode) => applyXFO.bind({ 'x-frame-options': mode })
+const applyXFO = (mode) => applyHeaders({ 'x-frame-options': mode })
+
+// {% if csrf %}
+// csrf protection middleware
+function signCookie(value, secret) {
+  return `${value}.${crypto.createHmac('sha256', secret).update(value).digest('base64')}`
+}
+
+function checkCookieSignature(input, secret) {
+  if (!input) {
+    return false
+  }
+  const [ message, signature ] = input.split('.', 2)
+  const valid = signCookie(message, secret)
+  return crypto.timingSafeEqual(Buffer.from(input), Buffer.from(valid)) ? message : false
```

Comment:
Notably, `crypto.timingSafeEqual` throws an exception if the two buffers are not of equal length, so we may want to bail early if `input.length != valid.length`.